core/src/backend/video.rs

```diff
@@ -0,0 +1,159 @@
+//! Video decoder backends
+
+use crate::backend::render::{BitmapInfo, RenderBackend};
+use generational_arena::{Arena, Index};
+use swf::{VideoCodec, VideoDeblocking};
+
+mod software;
+
+pub use crate::backend::video::software::SoftwareVideoBackend;
+
+pub type VideoStreamHandle = Index;
+
+pub type Error = Box<dyn std::error::Error>;
+
+/// An encoded video frame of some video codec.
+pub struct EncodedFrame<'a> {
+    /// The codec used to encode the frame.
+    pub codec: VideoCodec,
+
+    /// The raw bitstream data to funnel into the codec.
+    pub data: &'a [u8],
+
+    /// A caller-specified frame ID. Frame IDs must be consistent between
+    /// subsequent uses of the same data stream.
+    pub frame_id: u32,
+}
+
+impl<'a> EncodedFrame<'a> {
+    /// Borrow this frame's data.
+    pub fn data(&'a self) -> &'a [u8] {
+        &self.data
+    }
+}
+
+/// What dependencies a given video frame has on any previous frames.
+pub enum FrameDependency {
+    /// This frame has no reference frames and can be seeked to at any time.
+    Keyframe,
+
+    /// This frame has one reference frame which must be completely decoded
+    /// before this frame can be decoded.
+    ///
+    /// The reference frame is the previous frame in the video stream.
+    LastFrame,
+}
+
+/// A backend that provides access to some number of video decoders.
+///
+/// Implementations of `VideoBackend` are not required to actually support
+/// decoding any video formats. However, they must interoperate with at least
+/// one `RenderBackend` so that renderable video frames may be passed from the
+/// decoder to the renderer.
+pub trait VideoBackend {
+    /// Register a new video stream.
+    ///
+    /// Most of the parameters provided to this function are advisory: the
+    /// actual video data stream provided to the decoder may vary in size or
+    /// number of frames. This function should return an `Error` if it is not
```

Comment:
This is more like a question:
Is it likely that the deblocking filter mode is also just advisory, and the _actual_ setting to use is also embedded into the encoded data? Or is this something that has to be always taken into account, and performed after the decoder did its thing?

Reply:
In the case of H.263, picture size is indicated by `PTYPE` and/or `CPFMT`. Containing formats such as `SWF` will also state the picture size in `DefineVideoStream`. Since this is redundant, they can vary, and the comment here effectively means that the H.263 bitstream wins out. The decoded bitmap you get back from `decode_video_stream_frame` will be sized according to the bitstream size, not the SWF size.

I have not tried sending such an invalid movie through Flash Player to see what happens - I suspect it would either do what `video-h263` does, or crash horribly.

Reply:
This does not really answer my question about the deblocking filter mode, but I think I'll just generalise your "this is redundant" and "bitstream wins out" remarks to that as well, and won't worry much about simply ignoring what `DefineVideoStream` says about it for VP6. :)

Reply:
But now I wonder... What could be the point of repeating this value in the tag header?
For the frame size, I can imagine that it is useful for a number of reasons, and it's handy if the decoder doesn't have to be always involved in figuring it out...
But for deblocking??..

Reply:
Sorry, I misinterpreted your question. As for the filter specifically, I'm not sure, and I would imagine all sorts of wonky behaviors happen if the deblocking filter is incorrectly specified in this way.

For the record, I never implemented the Sorenson Spark deblocking filter in the H.263 decoder, so I didn't have to deal with that particular issue.

Reply:
I see. Thanks!